### PR TITLE
[UID2-3105] Update python SDK guide with new interface for DSP check

### PR DESCRIPTION
**Pull request by asloobq**
changes from https://github.com/IABTechLab/uid2-client-python/pull/31

### CODE REVIEW

File: docs/sdks/uid2-sdk-ref-python.md [hunk 1]
```diff
@@ -67,53 +67,145 @@ The package is published in this location:
 
 ## Initialization
 
-The initialization function configures the parameters necessary for the SDK to authenticate with the UID2 service. It also allows you to configure retry intervals in the event of errors.
+- DSPs should create an instance of the BidstreamClient class
```
Comment (mcollins-ttd) [inline docs/sdks/uid2-sdk-ref-python.md:70]: I think we should format class names as `code`

File: docs/sdks/uid2-sdk-ref-python.md [hunk 2]
```diff
@@ -67,53 +67,145 @@ The package is published in this location:
...
+- Sharers should create an instance of the SharingClient class
+- Publishers should create an instance of the Uid2PublisherClient class
 
-| Parameter | Description | Recommended Value |
-| :--- | :--- | :--- |
-| `endpoint` | The endpoint for the UID2 service. | N/A |
-| `authKey` | The authentication token that belongs to the client. For access to UID2, see [Contact Info](../getting-started/gs-account-setup.md#contact-info). | N/A |
+You will need to provide the values necessary for the SDK to authenticate with the UID2 service
```
Comment (mcollins-ttd) [inline docs/sdks/uid2-sdk-ref-python.md:74]: Full stop

File: docs/sdks/uid2-sdk-ref-python.md [hunk 3]
```diff
@@ -67,53 +67,145 @@ The package is published in this location:
...
+
+| Parameter    | Description                                                                               |
+|:-------------|:------------------------------------------------------------------------------------------|
+| `endpoint`   | The endpoint for the UID2 service. See [Environments](../getting-started/gs-environments) |
```
Comment (mcollins-ttd) [inline docs/sdks/uid2-sdk-ref-python.md:78]: Called `base_url`

File: docs/sdks/uid2-sdk-ref-python.md
Comment (mcollins-ttd) [inline docs/sdks/uid2-sdk-ref-python.md:1]: I removed the Overview section from the .NET SDK page, as it says that parameters and property names are pseducode. I suggest we do the same here.

https://github.com/IABTechLab/uid2docs/pull/490/files#diff-e343f1679a96ab6eeca175690e93e8a45e8b869ba667e5de5e939da7a41fbbefL26-L29

File: docs/sdks/uid2-sdk-ref-python.md (hunk 3 above)
Comment (asloobq) [reply docs/sdks/uid2-sdk-ref-python.md:78]: I was trying to sync this with the C# SDK and thus renamed it to endpoint. https://unifiedid.com/docs/sdks/uid2-sdk-ref-csharp-dotnet#initialization

File: docs/sdks/uid2-sdk-ref-python.md [hunk 4]
```diff
@@ -67,57 +63,149 @@ The package is published in this location:
...
+| Value                      | Description                                                             |
+|:---------------------------|:------------------------------------------------------------------------|
+| `SUCCESS`                  | The UID2 token was decrypted successfully and a raw UID2 was returned.  |
+| `NOT_AUTHORIZED_FOR_KEY`   | The requester does not have authorization to decrypt this UID2 token.   |
+| `NOT_INITIALIZED`          | The client library is waiting to be initialized.                        |
+| `INVALID_PAYLOAD`          | The incoming UID2 token is not a valid payload.                         |
+| `EXPIRED_TOKEN`            | The incoming UID2 token has expired.                                    |
+| `KEYS_NOT_SYNCED`          | The client has failed to synchronize keys from the UID2 service.        |
+| `VERSION_NOT_SUPPORTED`    | The client library does not support the version of the encrypted token. |
+| `DOMAIN_NAME_CHECK_FAILED` | The domain name doesn't match the the domain of the encrypted token.    |
```
Comment (genwhittTTD) [inline docs/sdks/uid2-sdk-ref-python.md:129]: the the > the

File: docs/sdks/uid2-sdk-ref-python.md [hunk 5]
```diff
@@ -67,57 +63,149 @@ The package is published in this location:
 
 ## Initialization
 
-The initialization function configures the parameters necessary for the SDK to authenticate with the UID2 service. It also allows you to configure retry intervals in the event of errors.
+- DSPs should create an instance of the `BidstreamClient` class.
```
Comment (genwhittTTD) [inline docs/sdks/uid2-sdk-ref-python.md:66]: Needs an intro line between the heading and the bullets. For example:
The initialization step depends on your role, as follows:
Something like that to introduce it. 

Also, don't say "should" -- it's just what they have to do. Suggest making this into a table instead. Here is a suggestion.

The initialization step depends on the role, as shown in the following table.

| For this role... | To initialize, create an instance of this class... |
| :--- | :--- |
| DSP | `BidstreamClient` |
| Sharer | `SharingClient` |
| Publisher | `Uid2PublisherClient` |

(iupdating based on later reading) One other suggestion is to add a link to the section with specific instructions for the role. Like this:

The initialization step depends on the role, as shown in the following table.

| Role | Create Instance of Class | Link to Instructions |
| :--- | :--- | :--- |
| DSP | `BidstreamClient` | [Usage for DSPs](#usage-for-dsps) |
| Sharer | `SharingClient` | [Usage for Sharers](#usage-for-uid2-sharers) |
| Publisher | `Uid2PublisherClient` | [Usage for Publishers](#usage-for-publishers) |

File: docs/sdks/uid2-sdk-ref-python.md [hunk 6]
```diff
@@ -67,57 +63,149 @@ The package is published in this location:
...
+
+The `SharingClient` class allows you to encrypt raw UID2s into UID2 tokens and decrypt UID2 tokens into raw UID2s.
+
 
 >NOTE: When you use an SDK, you do not need to store or manage decryption keys.
 
-If you're a DSP, for bidding, call the interface to decrypt a UID2 advertising token and return the UID2. For details on the bidding logic for handling user opt-outs, see [DSP Integration Guide](../guides/dsp-guide.md).
+### Encryption Response Content
+
+When encrypting with the `SharingClient`, the SDK returns the following information:
```
Comment (genwhittTTD) [inline docs/sdks/uid2-sdk-ref-python.md:90]: When encrypting with the `SharingClient`, the SDK returns the information shown in the following table.
(style guidelines, we don't use a colon before a table, only before steps or bullets).

File: docs/sdks/uid2-sdk-ref-python.md [hunk 7]
```diff
@@ -67,57 +63,149 @@ The package is published in this location:
...
+
+| Property         | Description                                                                                                                                     |
+|:-----------------|:------------------------------------------------------------------------------------------------------------------------------------------------|
+| `status`         | The encryption result status. For a list of possible values and definitions, see [Encryption Response Statuses](#encryption-response-statuses). |
+| `encrypted_data` | The encrypted UID2 token.                                                                                                                       |
+
+### Encryption Response Statuses
+
+| Value                           | Description                                                            |
```
Comment (genwhittTTD) [inline docs/sdks/uid2-sdk-ref-python.md:99]: Need some sort of intro line before the table. For example:

Encryption response codes, and their meanings, are shown in the following table.

I know it doesn't really add anything but it's a general rule for our docs.

File: docs/sdks/uid2-sdk-ref-python.md [hunk 8]
```diff
@@ -67,57 +63,149 @@ The package is published in this location:
...
+| `SUCCESS`                       | The raw UID2 was successfully encrypted and a UID2 token was returned. |
+| `NOT_AUTHORIZED_FOR_KEY`        | The requester does not have authorization to use the encryption key.   |
+| `NOT_AUTHORIZED_FOR_MASTER_KEY` | The requester does not have authorization to use the master key.       |
+| `NOT_INITIALIZED`               | The client library is waiting to be initialized.                       |
+| `KEYS_NOT_SYNCED`               | The client has failed to synchronize keys from the UID2 service.       |
+| `ENCRYPTION_FAILURE`            | A generic encryption failure occurred.                                 |
+
+### Decryption Response Content
+
+Whether decrypting with the `BidstreamClient` or the `SharingClient`, the SDK returns the following information:
```
Comment (genwhittTTD) [inline docs/sdks/uid2-sdk-ref-python.md:110]: ...the SDK returns the information shown in the following table.

**Review by genwhittTTD — COMMENTED** [review]
Some copy mods. I tried to make it easy. LMK if you have any questions.

Note: That heading "Usage for UID2 Sharers" -- we'll have to change that to "Usage for Sharers" -- to match the other two headings, and for portability to EUID. But there are four server-side SDKs, and I have lots of docs that link in to them also. So I didn't suggest this change. I'll do it all in one pass later.

File: docs/sdks/uid2-sdk-ref-python.md [hunk 9]
```diff
@@ -67,57 +63,149 @@ The package is published in this location:
...
 ## Interface 
 
-The interface allows you to decrypt UID2 advertising tokens and return the corresponding raw UID2. 
+The `BidstreamClient` class allows you to decrypt UID2 tokens into raw UID2s.
+For details on the bidding logic for handling user opt-outs, see [DSP Integration Guide](../guides/dsp-guide.md).
+
+The `SharingClient` class allows you to encrypt raw UID2s into UID2 tokens and decrypt UID2 tokens into raw UID2s.
+
 
 >NOTE: When you use an SDK, you do not need to store or manage decryption keys.
```
Comment (caroline-ttd) [inline docs/sdks/uid2-sdk-ref-python.md:86]:
```
:::note
When you use an SDK, you do not need to store or manage decryption keys.
:::

```
We can use `:::` to align with C# sdk as note

File: docs/sdks/uid2-sdk-ref-python.md [hunk 10]
```diff
@@ -67,57 +63,149 @@ The package is published in this location:
...
+
+| Property      | Description                                                                                                                                     |
+|:--------------|:------------------------------------------------------------------------------------------------------------------------------------------------|
+| `status`      | The decryption result status. For a list of possible values and definitions, see [Decryption Response Statuses](#decryption-response-statuses). |
+| `uid`         | The raw UID2 for the corresponding UID2 token.                                                                                                  |
+| `established` | The timestamp indicating when a user first established the UID2 with the publisher.                                                             |
+
+### Decryption Response Statuses
+
+| Value                      | Description                                                             |
```
Comment (caroline-ttd) [inline docs/sdks/uid2-sdk-ref-python.md:120]: Decryption response codes, and their meanings, are shown in the following table.

File: docs/sdks/uid2-sdk-ref-python.md [hunk 11]
```diff
@@ -67,57 +63,149 @@ The package is published in this location:
...
+|:---------------------------|:------------------------------------------------------------------------|
+| `SUCCESS`                  | The UID2 token was decrypted successfully and a raw UID2 was returned.  |
+| `NOT_AUTHORIZED_FOR_KEY`   | The requester does not have authorization to decrypt this UID2 token.   |
+| `NOT_INITIALIZED`          | The client library is waiting to be initialized.                        |
+| `INVALID_PAYLOAD`          | The incoming UID2 token is not a valid payload.                         |
+| `EXPIRED_TOKEN`            | The incoming UID2 token has expired.                                    |
+| `KEYS_NOT_SYNCED`          | The client has failed to synchronize keys from the UID2 service.        |
+| `VERSION_NOT_SUPPORTED`    | The client library does not support the version of the encrypted token. |
+| `DOMAIN_NAME_CHECK_FAILED` | The domain name doesn't match the the domain of the encrypted token.    |
+| `INVALID_TOKEN_LIFETIME`   | The lifetime of the incoming UID2 token is longer than expected.        |
```
Comment (caroline-ttd) [inline docs/sdks/uid2-sdk-ref-python.md:130]: Will "The token has invalid timestamp" be more accurate?

We may throw this error in two conditions: 1. token lifetime is longer than maxLifetime 2.skewDuration is longer than allow_clock_skew_seconds

File: docs/sdks/uid2-sdk-ref-python.md (hunk 9 above)
Comment (caroline-ttd) [reply docs/sdks/uid2-sdk-ref-python.md:86]: nvm, after reading the doc, I feel it does not matter.

File: docs/sdks/uid2-sdk-ref-python.md [hunk 12]
```diff
@@ -67,57 +63,149 @@ The package is published in this location:
...
+client = BidstreamClient(UID2_BASE_URL, UID2_API_KEY, UID2_SECRET_KEY)
+```
+
+2. Refresh once at startup, and then periodically (recommended refresh interval is hourly):
+
+```py
+client.refresh()
+```
+
+3. Decrypt a token into a raw UID2. Pass the domain name of the site where the bid originated from:
```
Comment (sunnywu) [inline docs/sdks/uid2-sdk-ref-python.md:148]: might wanna add this new edit from c# sdk https://github.com/IABTechLab/uid2docs/pull/518/files

File: docs/sdks/uid2-sdk-ref-python.md [hunk 13]
```diff
@@ -67,57 +63,149 @@ The package is published in this location:
...
+# If decryption succeeded, use the raw UID2.
+if decrypted.success:
+    #  Use decrypted.uid
+else:
+   # Check decrypted.status for the failure reason.
+```
+
+For a full example, see the `sample_bidstream_client.py` in [examples/sample_bidstream_client.py](https://github.com/IABTechLab/uid2-client-python/blob/main/examples/sample_bidstream_client.py).
+
+## Usage for UID2 Sharers
```
Comment (sunnywu) [inline docs/sdks/uid2-sdk-ref-python.md:161]: could we keep the Sharers section at the bottom like before? I'm not sure what the rationale is? But usage for Publishers definitely should be higher than Sharer for now.

File: docs/sdks/uid2-sdk-ref-python.md (hunk 5 above)
Comment (asloobq) [reply docs/sdks/uid2-sdk-ref-python.md:66]: thanks. That looks better!

File: docs/sdks/uid2-sdk-ref-python.md (hunk 11 above)
Comment (asloobq) [reply docs/sdks/uid2-sdk-ref-python.md:130]: Hmm. I had forgotten the second scenario. `invalid timestamp` is more generic

File: docs/sdks/uid2-sdk-ref-python.md (hunk 12 above)
Comment (asloobq) [reply docs/sdks/uid2-sdk-ref-python.md:148]: definitely!

File: docs/sdks/uid2-sdk-ref-python.md (hunk 13 above)
Comment (asloobq) [reply docs/sdks/uid2-sdk-ref-python.md:161]: Hmm. Will move this

File: docs/sdks/uid2-sdk-ref-python.md [hunk 14]
```diff
@@ -192,48 +242,48 @@ If you're using server-only integration (see [Publisher Integration Guide, Serve
 
 ## Usage for UID2 Sharers
 
-A UID2 sharer is any participant that wants to share UID2s with another participant. Raw UID2s must be encrypted into UID2 tokens before sending them to another participant. For an example of usage, see [examples/sample_sharing.py](https://github.com/IABTechLab/uid2-client-python/blob/master/examples/sample_sharing.py) script.
+A UID2 sharer is any participant that wants to share UID2s with another participant. Raw UID2s must be encrypted into UID2 tokens before sending them to another participant.
```
Comment (genwhittTTD) [inline docs/sdks/uid2-sdk-ref-python.md:245]: I have a question in to KL on this... may be an edit maybe not but noting it.

File: docs/sdks/uid2-sdk-ref-python.md [hunk 15]
```diff
@@ -66,58 +63,111 @@ The package is published in this location:
...
 
-| Parameter | Description | Recommended Value |
-| :--- | :--- | :--- |
-| `endpoint` | The endpoint for the UID2 service. | N/A |
-| `authKey` | The authentication token that belongs to the client. For access to UID2, see [Contact Info](../getting-started/gs-account-setup.md#contact-info). | N/A |
+You will need to provide the values necessary for the SDK to authenticate with the UID2 service.
+
+| Parameter    | Description                                                                               |
+|:-------------|:------------------------------------------------------------------------------------------|
+| `endpoint`   | The endpoint for the UID2 service. See [Environments](../getting-started/gs-environments) |
```
Comment (genwhittTTD) [inline docs/sdks/uid2-sdk-ref-python.md:79]: Line #79 -- just add a period at the end:

| `endpoint` | The endpoint for the UID2 service. See [Environments](../getting-started/gs-environments). |

File: docs/sdks/uid2-sdk-ref-python.md [hunk 16]
```diff
@@ -192,48 +242,48 @@ If you're using server-only integration (see [Publisher Integration Guide, Serve
...
-         # send encrypted_data to receiver
-      except EncryptionError as err:
-        # check for failure reason
-        print(err)
-      ```
+```py
+encrypted = client.encrypt_raw_uid_into_token(raw_uid)
+# If encryption succeeded, send the UID2 token to the receiver.
+if encrypted.success:
+    # Send encrypted.uid to receiver
```
Comment (caroline-ttd) [inline docs/sdks/uid2-sdk-ref-python.md:271]: In java, we send the `encrypted.getEncryptedData()` not uid, is it the same as `encrypted.uid` here?

Comment (asloobq) [reply docs/sdks/uid2-sdk-ref-python.md:271]: you are right, it should be `encrypted.encrypted_data`. Thanks for catching this

**Review by jon8787 — APPROVED** [review]

**Review by caroline-ttd — APPROVED** [review]
LGTM

File: docs/sdks/uid2-sdk-ref-python.md [hunk 17]
```diff
@@ -68,58 +65,114 @@ The package is published in this location:
...
+| `NOT_AUTHORIZED_FOR_KEY`   | The requester does not have authorization to decrypt this UID2 token.   |
+| `NOT_INITIALIZED`          | The client library is waiting to be initialized.                        |
+| `INVALID_PAYLOAD`          | The incoming UID2 token is not a valid payload.                         |
+| `EXPIRED_TOKEN`            | The incoming UID2 token has expired.                                    |
+| `KEYS_NOT_SYNCED`          | The client has failed to synchronize keys from the UID2 service.        |
+| `VERSION_NOT_SUPPORTED`    | The client library does not support the version of the encrypted token. |
+| `DOMAIN_NAME_CHECK_FAILED` | The domain name doesn't match the domain of the encrypted token.        |
+| `INVALID_TOKEN_LIFETIME`   | The token has an invalid timestamp.                                     |
+
+## Usage for DSPs
```
Comment (sunnywu) [inline docs/sdks/uid2-sdk-ref-python.md:141]: i think per @jon8787 comment [here ](https://thetradedesk.slack.com/archives/C051ELQ8JAC/p1714458862861149?thread_ts=1714439519.989579&cid=C051ELQ8JAC), could you move DSP section below Publisher's?

**Review by sunnywu — APPROVED** [review]

**Pull request merged**